### PR TITLE
Task/enum

### LINKED ISSUE #151: Write enum meta replacement
EnumMeta in the standard lib is slow, and during deserialization we invoke this a lot.

We should have a custom impl such that;

```py
if TYPE_CHECKING:
    EnumMeta = enum.EnumMeta
    Enum = enum.Enum
    ...
else:
    class EnumMeta(...):
        ...

    class Enum(...):
        ...
```
These should have a similar interface but be optimised where possible for speed.

Benchmarking is a must.

### PR DESCRIPTION
**Pull request by nekokatt**
FIxes #151, reimplements enum to be faster.

```
BasicPyEnum.__call__('25') 0.43657039100071415 µs
BasicHikariEnum.__call__('25') 0.19885178900040046 µs
BasicPyEnum._value2member_map_['25'] 0.09891081800014945 µs
BasicHikariEnum._value2member_map['25'] 0.1033127640002931 µs
BasicPyEnum.__getitem__['z'] 0.16736914299872296 µs
BasicHikariEnum.__getitem__['z'] 0.15685381999901438 µs
BasicPyEnum.__call__ profile
         2000003 function calls in 0.744 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.249    0.249    0.744    0.744 <string>:1(<module>)
  1000000    0.274    0.000    0.495    0.000 enum.py:283(__call__)
  1000000    0.221    0.000    0.221    0.000 enum.py:562(__new__)
        1    0.000    0.000    0.744    0.744 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


BasicHikariEnum.__call__ profile
         1000003 function calls in 0.340 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.183    0.183    0.340    0.340 <string>:1(<module>)
  1000000    0.157    0.000    0.157    0.000 enums.py:115(__call__)
        1    0.000    0.000    0.340    0.340 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


BasicPyEnum.__getitem__ profile
         1000003 function calls in 0.331 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.165    0.165    0.331    0.331 <string>:1(<module>)
  1000000    0.166    0.000    0.166    0.000 enum.py:348(__getitem__)
        1    0.000    0.000    0.331    0.331 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


BasicHikariEnum.__getitem__ profile
         1000003 function calls in 0.318 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.159    0.159    0.318    0.318 <string>:1(<module>)
  1000000    0.158    0.000    0.158    0.000 enums.py:141(__getitem__)
        1    0.000    0.000    0.318    0.318 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

### CODE REVIEW

File: hikari/utilities/enums.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Implementation of parts of Python's `enum` protocol to be faster."""
+from __future__ import annotations
+
+__all__: typing.List[str] = ["Enum"]
```
Comment (davfsa) [inline hikari/utilities/enums.py:25]: As I was reviewing this, I realized that we dont really have a style for this. Most of the lib uses `typing.Final[typing.List[str]]` but I would be willing to switch it to `typing.List[str]` only :smile:

File: hikari/utilities/enums.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+import os
+import sys
+import types
+import typing
+
+_T = typing.TypeVar("_T")
+
+
+class _EnumNamespace(typing.Dict[str, typing.Any]):
```
Comment (davfsa) [inline hikari/utilities/enums.py:35]: Slots?

**Review by davfsa — COMMENTED** [review]
LGTM, just two things :smile:

File: hikari/utilities/enums.py (hunk 2 above)
Comment (nekokatt) [reply hikari/utilities/enums.py:35]: Why does it need slots, it is used once per program, and in like 20 objects...

If you are having to microoptimise to that level, you shouldnt be using Python.

File: hikari/utilities/enums.py (hunk 1 above)
Comment (nekokatt) [reply hikari/utilities/enums.py:25]: Mistake, I'll fix it at some point when I use a computer.

**Pull request closed without merging**